Polymer 3 Conversion

qwangd2l opened the pull request.

qwangd2l commented on demo/index.html:
```diff
@@ -126,78 +155,86 @@ <h3>Hierarchical View</h3>
+document.getElementById('btn-view-4').addEventListener('click', function() {
+	showSubView('view4');
+});
+document.getElementById('btn-parent-view-2a').addEventListener('click', function() {
```
This event listener is new because the original code:
```
<button onclick="showParentView('view2a');">view 1 (parent)</button>
```
does not reference the function correctly

awikkerink replied:
So this was an existing bug in the hybrid branch? Could this be fixed there first?

qwangd2l replied:
Sorry, I meant it doesn't work after the conversion to polymer 3.
It works fine in the hybrid.

awikkerink replied:
Ah, I see. `showParentView` was global before. Gotcha

dbatiste replied:
In other project I also switched the globals like this to use `addEventListener`.  I would not be opposed to adding an exception to our testing eslint config to allow globals, and also console statements.

awikkerink commented:
I've seen this before. I think it's caused by the <custom-style> tags being in the <head> rather than the body

qwangd2l commented:
> I've seen this before. I think it's caused by the tags being in the rather than the body
> 
> --edit no, `document.body.appendChild` is used. Hmm.
> --edit So, I think before the styling wasn't getting applied correctly, but it is now fixed after the conversion. 🎉

Hmm, I should mention that I added the following to the hybrid (only locally) to get d2l-typography to apply.
```
<script>
	document.body.classList.add('d2l-typography');
</script>
``` (after </body>)
Then I compared its demo with the one after converting to polymer 3. The differences I'm seeing are described in my earlier comment (different font sizes and container size - although not too different).

awikkerink commented:
Chrome does weird things if you add unrecognised tags to the head element. I think it moves the custom-style elements to body, and by doing so it ignores the attributes in the body tag.

qwangd2l commented:
Here are some screenshots to illustrate the differences
hybrid:
![hierarchical_demo_hybrid](https://user-images.githubusercontent.com/37313128/48375697-b28f0380-e696-11e8-95fd-d2a31308a8f8.png)
polymer 3:
![hierarchical_demo_polymer3](https://user-images.githubusercontent.com/37313128/48375698-b28f0380-e696-11e8-87bc-26b73e5c94ca.png)

dbatiste commented on d2l-hierarchical-view.js:
```diff
@@ -0,0 +1,40 @@
+			<slot></slot>
+		</div>
+	</template>
+
```
Looks like unnecessary whitespace here.  This can probably be cleaned up in the hybrid beforehand.

qwangd2l replied:
I can remove the unnecessary whitespace here for this PR.
Would this be ok to do or should I make a PR for the hybrid and re-run the conversion process?

dbatiste replied:
Suggesting to rerun the conversion.  Not a big deal if linting passes.  Usually linter errors on extra whitespace but this is inside of a template so it may not.  If linting passes I wouldn't worry about it unless you are doing a hybrid PR anyway.

qwangd2l commented on demo/index.html:
```diff
@@ -126,78 +155,86 @@ <h3>Hierarchical View</h3>
+		button.focus();
+	}
+};
+var rootView = document.getElementById('view1');
```
This doesn't seem like an easy fix. The WebComponentsReady event doesn't seem to get fired within a `<script type="module">` tag. I've also tried adding a separate script tag to listen for `WebComponentsReady` but `document.getElementById(...)` returns null
I've also noticed that most of the buttons don't do anything in firefox in the hybrid version (because `document.getElementById(...)` returns null) - same thing seems to happen with safari.

dbatiste replied:
The buttons not working in FF sounds odd - they definitely used to, and I don't think much has changed.  I think should solve that before trying to convert.  I think it should be possible for all this wire up to happen regardless of WCR.

qwangd2l replied:
So I found that the document ready state was `interactive` and didn't allow `document.getElementById` to get anything until the state was `complete` (in Firefox anyway).
The following seems to work in Firefox (as well as Safari and Chrome)
```
document.addEventListener('readystatechange', event => {
  if (event.target.readyState === "complete") {
     <insert demo code here to render buttons and attach event listeners to buttons>
  }
});
```
Any thoughts on if this is what we should do?

dbatiste commented:
Note: I'm not spotting any differences in those screenshots either.  If there is a font issue we should get to the bottom of that to make sure we don't have issues with it elsewhere.  

Aside, I think the typography was only really being includes so that the demo page bares some resemblance to the rest of our demo pages.  If it makes things easier, I think removing `d2l-typography` from this demo may be possible.

qwangd2l commented:
> Note: I'm not spotting any differences in those screenshots either. If there is a font issue we should get to the bottom of that to make sure we don't have issues with it elsewhere.
> 
> Aside, I think the typography was only really being includes so that the demo page bares some resemblance to the rest of our demo pages. If it makes things easier, I think removing `d2l-typography` from this demo may be possible.

When I inspect the hybrid demo and the polymer 3 demo, there doesn't seem to be any differences in font size or font. But the number of lines to render the same text is different, as well as the content container being different height (224px in polymer 3 demo, 196px in hybrid demo).

dbatiste commented:
That is interesting.  Is there a difference in height between hybrid using shady and shadow-DOM?

qwangd2l commented:
> That is interesting. Is there a difference in height between hybrid using shady and shadow-DOM?

The height doesn't seem to change in the hybrid demo when using shadow dom.
I'm enabling shadow dom through the query string: `/components/d2l-hierarchical-view/demo/index.html?dom=shadow`
I've also tried enabling shadow dom through adding the following before `webcomponents-lite.js` import
```
<script>
window.Polymer = {
	dom: 'shadow',
};
</script>
```

qwangd2l commented:
Going to close this pull request because there's another PR with a fix (which also includes Polymer 3 conversion) to the WebComponentsReady question/problem:

https://github.com/BrightspaceUI/hierarchical-view/pull/34
